src/ir/module-splitting.cpp

```diff
@@ -0,0 +1,405 @@
+      assert(false && "TODO: handle ref.func as well");
+    }
+  };
+  PassRunner runner(&primary);
+  CallIndirector(
+    secondaryFuncs, secondarySignatures, getIndex, Builder(primary))
+    .run(&runner, &primary);
+
+  // Insert new table elements
+  if (newTableElems.size()) {
+    primary.table.exists = true;
+    secondary.table.exists = true;
+    // Update table sizes if necessary
+    size_t tableSize = firstFreeIndex + newTableElems.size();
+    if (primary.table.initial < tableSize) {
+      primary.table.initial = tableSize;
+      secondary.table.initial = tableSize;
+    }
+    if (primary.table.max < tableSize) {
+      primary.table.max = tableSize;
+      secondary.table.max = tableSize;
+    }
+    if (lastSegment != nullptr) {
+      lastSegment->data.insert(
+        lastSegment->data.end(), newTableElems.begin(), newTableElems.end());
+    } else {
+      primary.table.segments.emplace_back(
+        builder.makeConst(Literal(int32_t(firstFreeIndex))), newTableElems);
+    }
+  }
+}
+
+void exportImportPrimaryFunctions(Module& primary,
+                                  Module& secondary,
+                                  const std::set<Name>& primaryFuncs,
+                                  Name importNamespace) {
+  // Find primary functions called in the secondary module.
+  ModuleUtils::ParallelFunctionAnalysis<std::set<Name>> callCollector(
+    secondary, [&](Function* func, std::set<Name>& calledPrimaryFuncs) {
+      struct CallCollector : PostWalker<CallCollector> {
+        const std::set<Name>& primaryFuncs;
+        std::set<Name>& calledPrimaryFuncs;
+        CallCollector(const std::set<Name>& primaryFuncs,
+                      std::set<Name>& calledPrimaryFuncs)
+          : primaryFuncs(primaryFuncs), calledPrimaryFuncs(calledPrimaryFuncs) {
+        }
+
+        void visitCall(Call* curr) {
+          if (primaryFuncs.find(curr->target) != primaryFuncs.end()) {
+            calledPrimaryFuncs.insert(curr->target);
+          }
+        }
+      };
+      CallCollector(primaryFuncs, calledPrimaryFuncs).walkFunction(func);
+    });
+  std::set<Name> calledPrimaryFuncs;
+  for (auto& entry : callCollector.map) {
+    auto& calledFuncs = entry.second;
+    calledPrimaryFuncs.insert(calledFuncs.begin(), calledFuncs.end());
+  }
+
+  // Find exported primary functions and map to their export names
+  std::map<Name, Name> exportedPrimaryFuncs;
+  for (auto& ex : primary.exports) {
+    if (ex->kind == ExternalKind::Function) {
+      exportedPrimaryFuncs.insert(std::make_pair(ex->value, ex->name));
+    }
+  }
+
+  // Ensure each called primary function is exported and imported
+  for (auto primaryFunc : calledPrimaryFuncs) {
+    Name exportName;
+    auto exportIt = exportedPrimaryFuncs.find(primaryFunc);
+    if (exportIt != exportedPrimaryFuncs.end()) {
+      exportName = exportIt->second;
+    } else {
+      exportName = Names::getValidExportName(primary, primaryFunc);
```

Comment:
I wonder if we should use a unique / configurable prefix for the exports do distinguish between the actual function exports and the exports that are used for module splitting?   I'd even suggest that maybe we should ignore any existing exports and export always with our specially-named exports.

We could even pick a reasonable default mangling scheme that doesn't collide with C symbol namespace?  How about "%" as the prefix?  Or maybe that is too cute.

Reply:
Is the worry that users will accidentally depend on the new exports? I'm not sure how serious of a problem that would be. Even if we did use a special prefix, there's nothing stopping the same prefix being used in the input module, so I'm not sure this would buy us much. Maybe it would make the tests more readable?

Reply:
I was thinking for a tool like emscripten we might want to use a prefix (and it would be easy for us to choose a collision-free one) so that we could treat the exports of a wasm module as its public API.  i.e. we could inspect the wasm module at runtime an export the public functions out to the world of JS.   

It makes the actual API of the final wasm module self-describing.  Without this I can't tell what is the public API and what isn't, right?

Reply:
Ah, that seems reasonable. I'll add a configuration option for this.